### PR TITLE
Add script to get snapcraft releases

### PR DESCRIPTION
**Pull request by FirelightFlagboy**

### CODE REVIEW

File: misc/snapcraft_releases.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+
+class Track(StrEnum):
+    Latest = "latest"
+    V2 = "v2"
+    V3 = "v3"
+    Nightly = "nightly"
+
+
+class Channel(StrEnum):
```
Comment (mmmarcos) [inline misc/snapcraft_releases.py:20]: [According to the docs](https://snapcraft.io/docs/channels), these are not exactly channels, but `risk-levels`. I would suggest one of:
```suggestion
class ChannelLevel(StrEnum):
```
```suggestion
class RiskLevel(StrEnum):
```
```suggestion
class Level(StrEnum):
```

File: misc/snapcraft_releases.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    if version.prerelease:
+        for pre, chan in PRERELEASE_TO_CHANNEL.items():
+            if version.prerelease.startswith(pre):
+                return chan
+        raise ValueError(f"Unsupported prerelease: {version.prerelease}")
+    elif version.dev is not None:
+        return Channel.Edge
+    else:
+        return Channel.Stable
```
Comment (mmmarcos) [inline misc/snapcraft_releases.py:65]:
```suggestion
    if version.prerelease:
        PRERELEASE_TO_CHANNEL = {
            "a": Channel.Edge,
            "b": Channel.Beta,
            "rc": Channel.Candidate,
        }
        for pre, chan in PRERELEASE_TO_CHANNEL.items():
            if version.prerelease.startswith(pre):
                return chan
        raise ValueError(f"Unsupported prerelease: {version.prerelease}")
    
    if version.dev is not None:
        return Channel.Edge
    
    return Channel.Stable
```

File: misc/snapcraft_releases.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    tracks: list[Track] = []
+    match version.major:
+        case 2:
+            tracks.extend(TRACKS_FOR_V2)
+        case 3:
+            tracks.extend(TRACKS_FOR_V3)
+        case _:
+            raise ValueError(f"Unsupported major version: {version.major}")
+    return tracks
```
Comment (mmmarcos) [inline misc/snapcraft_releases.py:43]:
```suggestion
    match version.major:
        case 2:
            return (Track.V2,)
        case 3:
		    # TODO: Add Latest track once we publicly release v3
    		return (Track.V3,)
        case _:
            raise ValueError(f"Unsupported major version: {version.major}")
```

File: misc/snapcraft_releases.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+assert get_channel_for_version(Version.parse("3.0.0-rc.0")) == Channel.Candidate
+assert get_channel_for_version(Version.parse("3.0.0-dev.0")) == Channel.Edge
+
+if __name__ == "__main__":
+    parser = ArgumentParser(
+        "snapcraft_channel",
+        description="""
+    Determine the snapcraft channel based on the provided options and version.
+    """,
+    )
```
Comment (mmmarcos) [inline misc/snapcraft_releases.py:80]: Script is named `snapcraft_releases`, I would change descriptions here:
```suggestion
    parser = ArgumentParser(
        "snapcraft_releases",
        description="""
    Output as many snapcraft's `--release=<channel>` options based on the given Parsec version.
    """,
        epilog="See https://snapcraft.io/docs/channels")
    )
```

File: misc/snapcraft_releases.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    parser.add_argument(
+        "version", help="The version to determine the channel for.", type=Version.parse
```
Comment (mmmarcos) [inline misc/snapcraft_releases.py:83]:
```suggestion
        "version", help="The Parsec version to determine the channels for.", type=Version.parse
```

File: misc/snapcraft_releases.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    )
+    parser.add_argument("--nightly", help="Use nightly channel", action="store_true")
```
Comment (mmmarcos) [inline misc/snapcraft_releases.py:85]:
```suggestion
    parser.add_argument("--nightly", help="Use nightly track", action="store_true")
```

File: misc/snapcraft_releases.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    args = parser.parse_args()
+
+    logging.basicConfig(level=logging.INFO)
+
+    version: Version = args.version
+    log.info(f"Parsed version: {version!r}")
```
Comment (mmmarcos) [inline misc/snapcraft_releases.py:92]:
```suggestion
    log.info(f"Version: {version!r}")
```

File: misc/snapcraft_releases.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            *(map(lambda track: f"--release={track}/{Channel.Edge}", tracks_for_version)),
+            # Update all nightly's channels
+            *(map(lambda chan: f"--release={Track.Nightly}/{chan}", CHANNELS)),
+            sep="\n",
+        )
+    else:
+        channel = get_channel_for_version(version)
+        log.info(f"Channel for version {version}: {channel}")
+
+        print(*(map(lambda track: f"--release={track}/{channel}", tracks_for_version)), sep="\n")
```
Comment (mmmarcos) [inline misc/snapcraft_releases.py:109]: Wouldn't be more useful to directly log the output of the script?
```suggestion
    tracks_for_version = get_tracks_for_version(version)

    release_channels = []
    if args.nightly:
        # Update channels "<tracks>/edge"
        release_channels.extend(map(lambda track: f"--release={track}/{Channel.Edge}", tracks_for_version))
        # Update channels "nightly/<risk>"
        release_channels.extend(map(lambda chan: f"--release={Track.Nightly}/{chan}", CHANNELS))
    else:
        channel = get_channel_for_version(version)
        release_channels.extend(map(lambda track: f"--release={track}/{channel}", tracks_for_version))

log.info(" ".join(release_channels))
print(*release_channels, sep="\n")
```

File: misc/snapcraft_releases.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
```
Comment (mmmarcos) [inline misc/snapcraft_releases.py:1]:
```suggestion
# Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
r"""
For a given Parsec version, this script outputs the `--release=<channel>` options intended for snapcraft.

- A channel is formed with <track>/<risk-level>/<branch>
- Parsec currently has 4 tracks: latest, v2, v3, nigthly
- The <branch> component is not used

E.g.

$ python misc/snapcraft_releases.py 2.0.0 
--release=v2/stable
$ python misc/snapcraft_releases.py 3.0.0-b.0 
--release=v3/beta
--release=latest/beta
$ python misc/snapcraft_releases.py 3.0.0-b.0 --nightly 
--release=v3/edge
--release=latest/edge
--release=nightly/stable
--release=nightly/beta
--release=nightly/edge
--release=nightly/candidate
"""
```

**Review by mmmarcos — COMMENTED** [review]
A couple of suggestions

File: misc/snapcraft_releases.py (hunk 8 above)
Comment (FirelightFlagboy) [reply misc/snapcraft_releases.py]: done

**Review by mmmarcos — APPROVED** [review]
:+1:

**Pull request merged**